url72c32f3c54_Quickhull3d_quickhull3d_tgz-pJ8-com_github_quickhull3d_SimpleExampleJ8/src/com/github/quickhull3d/QuickHull3D.java

```diff
@@ -0,0 +1,1334 @@
+            do {
+                HalfEdge heOpp = findHalfEdge(he.head(), he.tail());
+                if (heOpp != null) {
+                    he.setOpposite(heOpp);
+                }
+                he = he.next;
+            } while (he != face.he0);
+            faces.add(face);
+        }
+    }
+
+    private void printQhullErrors(Process proc) throws IOException {
+        boolean wrote = false;
+        InputStream es = proc.getErrorStream();
+        StringBuffer error = new StringBuffer();
+        while (es.available() > 0) {
+            error.append((char) es.read());
+            wrote = true;
+        }
+        if (wrote) {
+            error.append(" ");
+            LOG.error(error.toString());
+        }
+    }
+
+    protected void setFromQhull(double[] coords, int nump, boolean triangulate) {
+        String commandStr = "./qhull i";
+        if (triangulate) {
+            commandStr += " -Qt";
+        }
+        try {
+            Process proc = Runtime.getRuntime().exec(commandStr);
+            PrintStream ps = new PrintStream(proc.getOutputStream(), false, Charset.defaultCharset().name());
+            StreamTokenizer stok = new StreamTokenizer(new InputStreamReader(proc.getInputStream(), Charset.defaultCharset()));
+
+            ps.println("3 " + nump);
+            for (int i = 0; i < nump; i++) {
+                ps.println(coords[i * 3 + 0] + " " + coords[i * 3 + 1] + " " + coords[i * 3 + 2]);
+            }
+            ps.flush();
+            ps.close();
+            Vector indexList = new Vector(3);
+            stok.eolIsSignificant(true);
+            printQhullErrors(proc);
+
+            do {
+                stok.nextToken();
+            } while (stok.sval == null || !stok.sval.startsWith("MERGEexact"));
+            for (int i = 0; i < 4; i++) {
+                stok.nextToken();
+            }
+            if (stok.ttype != StreamTokenizer.TT_NUMBER) {
+                throw new IllegalArgumentException("Expecting number of faces");
+            }
+            int numf = (int) stok.nval;
+            stok.nextToken(); // clear EOL
+            int[][] faceIndices = new int[numf][];
+            for (int i = 0; i < numf; i++) {
+                indexList.clear();
+                while (stok.nextToken() != StreamTokenizer.TT_EOL) {
+                    if (stok.ttype != StreamTokenizer.TT_NUMBER) {
+                        throw new IllegalArgumentException("Expecting face index");
+                    }
+                    indexList.add(0, Integer.valueOf((int) stok.nval));
+                }
+                faceIndices[i] = new int[indexList.size()];
+                int k = 0;
+                for (Iterator it = indexList.iterator(); it.hasNext();) {
+                    faceIndices[i][k++] = ((Integer) it.next()).intValue();
+                }
+            }
+            setHull(coords, nump, faceIndices, numf);
+        } catch (IllegalArgumentException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An exception is caught and re-thrown within the catch clause without any other operations being performed on the exception, making the catch clause unnecessary.

**Fix**: Perform additional operations on the exception that has been caught. You can wrap a standard exception into a custom exception to add additional information or special handling. When you do that, make sure to set the original exception as the cause. Otherwise, you will lose the stack trace and message of the original exception, which will make it difficult to analyze the event that caused the exception.  
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)